snapcraft/tests/fixture_setup.py

```diff
@@ -875,3 +877,57 @@ def fetch_binary(self, dir_, progress):
 
     def get_dependencies(self, _):
         return []
+
+
+class WithSnapNotInstalled(fixtures.Fixture):
+
+    def __init__(self, snap_name):
+        super().__init__()
+        self.snap_name = snap_name.split('/')[0]
+
+    def setUp(self):
+        super().setUp()
+        if snapcraft.repo.snaps.SnapPackage.is_snap_installed(self.snap_name):
+            raise AssertionError(
+                "This test cannot run if you already have the {snap!r} snap "
+                "installed. Please uninstall it by running "
+                "'sudo snap remove {snap}'.".format(snap=self.snap_name))
+
+        self.addCleanup(self._remove_snap)
+
+    def _remove_snap(self):
+        try:
+            subprocess.check_output(
+                ['sudo', 'snap', 'remove', self.snap_name],
+                stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            RuntimeError("unable to remove {!r}: {}".format(
+                self.snap_name, e.output))
+
+
+class SnapcraftYaml(fixtures.Fixture):
+
+    def __init__(
+            self, path, name='test-snap', version='test-version',
+            summary='test-summary', description='test-description'):
+        super().__init__()
+        self.path = path
+        self.data = {
+            'name': name,
+            'version': version,
+            'summary': summary,
+            'description': description,
+        }
+
+    def add_part(self, name, data):
+        part = {name: data}
+        if 'parts' in self.data:
```

Comment:
The smarts to update an existing part aren't obvious from the name. It would be clearer to my mind to have an explicit `extend_part` method.

Reply:
I would go the opposite path, no where in this code base is there a need to update a part and given it comes from a clean fixture we should raise an error if the part name is already added.

Reply:
Maybe I'm missing an edit here, but this doesn't handle the case of when a specific part already exists, but merely when _a_ part exists (in which case we want to add another, not blow away the one that already exists).

Reply:
I can't parse your phrasing... I'd expect `add_part("foo", {})` called twice to be an error. If there's a use case for modifying an existing part 'foo' it should be separate from this method.

Reply:
I've renamed this to update_part and copied the behaviour of a dict.